File: android/app/src/main/java/com/screentime/guardian/domain/ScreenTimeProvider.kt
package com.screentime.guardian.domain

import kotlinx.coroutines.flow.Flow
import java.time.LocalDate
import java.util.UUID

/**
 * Abstrahiert Android UsageStatsManager und iOS ScreenTime API
 */
interface ScreenTimeProvider {
    // Echtzeit-Daten
    suspend fun getCurrentSession(): AppSession?
    suspend fun getTodayUsage(): DailyUsage
    suspend fun getAppUsage(packageName: String): AppUsageStats?
    
    // Historie (30 Tage Local Cache)
    suspend fun getUsageHistory(days: Int = 30): List<DailyUsage>
    suspend fun getCategoryUsage(category: AppCategory, days: Int = 7): CategoryUsage
    
    // Limits
    suspend fun setAppLimit(packageName: String, limitMinutes: Int)
    suspend fun setCategoryLimit(category: AppCategory, limitMinutes: Int)
    suspend fun removeLimit(id: String)
    suspend fun getActiveLimits(): List<LimitEntity>
    
    // Events
    fun onLimitReached(): Flow<LimitEvent>
    fun onUnlock(): Flow<UnlockEvent>
}

// Data Models

data class AppSession(
    val id: String = UUID.randomUUID().toString(),
    val packageName: String,
    val appName: String,
    val category: AppCategory,
    val startTime: Long, // Unix timestamp (ms)
    val endTime: Long?, // null = currently active
    val durationSeconds: Long
) {
    val isActive: Boolean get() = endTime == null
}

data class DailyUsage(
    val date: LocalDate,
    val totalScreenTimeSeconds: Long,
    val unlockCount: Int,
    val appUsage: List<AppUsageBreakdown>,
    val categoryUsage: Map<AppCategory, Long>,
    val focusSessionsCompleted: Int,
    val focusTimeSeconds: Long
) {
    val id: String get() = date.toString() // YYYY-MM-DD
    val totalScreenTimeHours: Double get() = totalScreenTimeSeconds / 3600.0
    val focusTimeMinutes: Double get() = focusTimeSeconds / 60.0
}

data class AppUsageBreakdown(
    val packageName: String,
    val appName: String,
    val category: AppCategory,
    val usageSeconds: Long,
    val percentageOfTotal: Float
)

data class AppUsageStats(
    val packageName: String,
    val appName: String,
    val category: AppCategory,
    val totalUsageToday: Long,
    val lastUsed: Long?
)

data class CategoryUsage(
    val category: AppCategory,
    val totalSeconds: Long,
    val topApps: List<AppUsageBreakdown>
)

// Limit Models

data class LimitEntity(
    val id: String = UUID.randomUUID().toString(),
    val type: LimitType,
    val targetId: String, // packageName oder category.name
    val targetName: String,
    val limitMinutes: Int,
    val createdAt: Long = System.currentTimeMillis(),
    val isActive: Boolean = true,
    val currentUsageMinutes: Int? = null
)

enum class LimitType {
    APP, CATEGORY
}

// App Category (System-Kategorien)

enum class AppCategory(val displayName: String, val icon: String) {
    GAMES("Games", "videogame_asset"),
    SOCIAL("Social", "chat_bubble"),
    PRODUCTIVITY("Productivity", "check_circle"),
    ENTERTAINMENT("Entertainment", "play_circle"),
    EDUCATION("Education", "school"),
    HEALTH("Health", "favorite"),
    NEWS("News", "newspaper"),
    SHOPPING("Shopping", "shopping_bag"),
    FINANCE("Finance", "account_balance"),
    OTHER("Other", "apps");
    
    companion object {
        fun fromSystemCategory(category: Int): AppCategory {
            return when (category) {
                0 -> GAMES
                1 -> SOCIAL
                2 -> PRODUCTIVITY
                3 -> ENTERTAINMENT
                4 -> EDUCATION
                5 -> HEALTH
                6 -> NEWS
                7 -> SHOPPING
                8 -> FINANCE
                else -> OTHER
            }
        }
    }
}

// Events

data class LimitEvent(
    val limitId: String,
    val type: LimitType,
    val targetName: String,
    val limitMinutes: Int,
    val actualUsageMinutes: Int,
    val timestamp: Long = System.currentTimeMillis()
)

data class UnlockEvent(
    val limitId: String,
    val unlockedAt: Long = System.currentTimeMillis(),
    val unlockMethod: UnlockMethod
)

enum class UnlockMethod {
    BIOMETRIC, PIN, EMERGENCY
}

File: ios/ScreenTimeGuardian/Domain/ScreenTimeProvider.swift
//
//  ScreenTimeProvider.swift
//  ScreenTime Guardian - Domain Layer
//

import Foundation
import Combine

// MARK: - ScreenTimeProvider Protocol
/// Abstrahiert Android UsageStatsManager und iOS ScreenTime API
public protocol ScreenTimeProvider: ObservableObject {
    // MARK: - Real-time Data
    func getCurrentSession() async throws -> AppSession?
    func getTodayUsage() async throws -> DailyUsage
    func getAppUsage(bundleIdentifier: String) async throws -> AppUsageStats?
    
    // MARK: - History (30 Days Local Cache)
    func getUsageHistory(days: Int) async throws -> [DailyUsage]
    func getCategoryUsage(_ category: AppCategory, days: Int) async throws -> CategoryUsage
    
    // MARK: - Limits
    func setAppLimit(bundleIdentifier: String, limitMinutes: Int) async throws
    func setCategoryLimit(_ category: AppCategory, limitMinutes: Int) async throws
    func removeLimit(id: String) async throws
    func getActiveLimits() async throws -> [LimitEntity]
    
    // MARK: - Events
    var limitReachedPublisher: AnyPublisher<LimitEvent, Never> { get }
    var unlockPublisher: AnyPublisher<UnlockEvent, Never> { get }
}

// MARK: - Data Models

public struct AppSession: Identifiable, Codable, Equatable {
    public let id: UUID
    public let bundleIdentifier: String
    public let appName: String
    public let category: AppCategory
    public let startTime: Date
    public let endTime: Date?
    public let durationSeconds: TimeInterval
    
    public init(id: UUID = UUID(), 
                bundleIdentifier: String, 
                appName: String, 
                category: AppCategory, 
                startTime: Date, 
                endTime: Date?, 
                durationSeconds: TimeInterval) {
        self.id = id
        self.bundleIdentifier = bundleIdentifier
        self.appName = appName
        self.category = category
        self.startTime = startTime
        self.endTime = endTime
        self.durationSeconds = durationSeconds
    }
    
    public var isActive: Bool { endTime == nil }
}

public struct DailyUsage: Identifiable, Codable, Equatable {
    public let id: String // YYYY-MM-DD
    public let date: Date
    public let totalScreenTimeSeconds: TimeInterval
    public let unlockCount: Int
    public let appUsage: [AppUsageBreakdown]
    public let categoryUsage: [AppCategory: TimeInterval]
    public let focusSessionsCompleted: Int
    public let focusTimeSeconds: TimeInterval
    
    public var totalScreenTimeHours: Double {
        totalScreenTimeSeconds / 3600
    }
    
    public var focusTimeMinutes: Double {
        focusTimeSeconds / 60
    }
    
    public init(date: Date, 
                totalScreenTimeSeconds: TimeInterval,
                unlockCount: Int,
                appUsage: [AppUsageBreakdown],
                categoryUsage: [AppCategory: TimeInterval],
                focusSessionsCompleted: Int,
                focusTimeSeconds: TimeInterval) {
        self.id = date.formatted(.iso8601.year().month().day())
        self.date = date
        self.totalScreenTimeSeconds = totalScreenTimeSeconds
        self.unlockCount = unlockCount
        self.appUsage = appUsage
        self.categoryUsage = categoryUsage
        self.focusSessionsCompleted = focusSessionsCompleted
        self.focusTimeSeconds = focusTimeSeconds
    }
}

public struct AppUsageBreakdown: Codable, Equatable {
    public let bundleIdentifier: String
    public let appName: String
    public let category: AppCategory
    public let usageSeconds: TimeInterval
    public let percentageOfTotal: Float
}

public struct AppUsageStats: Codable, Equatable {
    public let bundleIdentifier: String
    public let appName: String
    public let category: AppCategory
    public let totalUsageToday: TimeInterval
    public let lastUsed: Date?
}

public struct CategoryUsage: Codable, Equatable {
    public let category: AppCategory
    public let totalSeconds: TimeInterval
    public let topApps: [AppUsageBreakdown]
}

// MARK: - Limit Models

public struct LimitEntity: Identifiable, Codable, Equatable {
    public let id: String
    public let type: LimitType
    public let targetId: String // bundleIdentifier oder category.rawValue
    public let targetName: String
    public let limitMinutes: Int
    public let createdAt: Date
    public var isActive: Bool
    public var currentUsageMinutes: Int?
    
    public init(id: String = UUID().uuidString,
                type: LimitType,
                targetId: String,
                targetName: String,
                limitMinutes: Int,
                createdAt: Date = Date(),
                isActive: Bool = true,
                currentUsageMinutes: Int? = nil) {
        self.id = id
        self.type = type
        self.targetId = targetId
        self.targetName = targetName
        self.limitMinutes = limitMinutes
        self.createdAt = createdAt
        self.isActive = isActive
        self.currentUsageMinutes = currentUsageMinutes
    }
}

public enum LimitType: String, Codable, Equatable {
    case app
    case category
}

// MARK: - App Category (System Categories)

public enum AppCategory: String, Codable, CaseIterable, Equatable {
    case games = "GAMES"
    case social = "SOCIAL"
    case productivity = "PRODUCTIVITY"
    case entertainment = "ENTERTAINMENT"
    case education = "EDUCATION"
    case health = "HEALTH"
    case news = "NEWS"
    case shopping = "SHOPPING"
    case finance = "FINANCE"
    case other = "OTHER"
    
    public var displayName: String {
        switch self {
        case .games: return "Games"
        case .social: return "Social"
        case .productivity: return "Productivity"
        case .entertainment: return "Entertainment"
        case .education: return "Education"
        case .health: return "Health"
        case .news: return "News"
        case .shopping: return "Shopping"
        case .finance: return "Finance"
        case .other: return "Other"
        }
    }
    
    public var icon: String {
        switch self {
        case .games: return "gamecontroller.fill"
        case .social: return "bubble.left.and.bubble.right.fill"
        case .productivity: return "checkmark.square.fill"
        case .entertainment: return "play.rectangle.fill"
        case .education: return "book.fill"
        case .health: return "heart.fill"
        case .news: return "newspaper.fill"
        case .shopping: return "bag.fill"
        case .finance: return "dollarsign.circle.fill"
        case .other: return "app.fill"
        }
    }
}

// MARK: - Events

public struct LimitEvent: Codable, Equatable {
    public let limitId: String
    public let type: LimitType
    public let targetName: String
    public let limitMinutes: Int
    public let actualUsageMinutes: Int
    public let timestamp: Date
}

public struct UnlockEvent: Codable, Equatable {
    public let limitId: String
    public let unlockedAt: Date
    public let unlockMethod: UnlockMethod
}

public enum UnlockMethod: String, Codable, Equatable {
    case biometric = "BIOMETRIC"
    case pin = "PIN"
    case emergency = "EMERGENCY"
}
